demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c

```diff
@@ -0,0 +1,675 @@
+        pucSig[ 3 ] = 0x21;
+        pucSig[ 4 ] = 0x0;
+        memcpy( &pucSig[ 5 ], ucTemp, 32 );
+        pucSigPtr = pucSig + 33 + 4;
+    }
+    else
+    {
+        pucSig[ 3 ] = 0x20;
+        memcpy( &pucSig[ 4 ], ucTemp, 32 );
+        pucSigPtr = pucSig + 32 + 4;
+    }
+
+    pucSigPtr[ 0 ] = 0x02; /* Integer. */
+    pucSigPtr++;
+
+    if( ucTemp[ 32 ] & 0x80 )
+    {
+        pucSig[ 1 ]++;
+        pucSigPtr[ 0 ] = 0x21;
+        pucSigPtr[ 1 ] = 0x00;
+        pucSigPtr += 2;
+
+        memcpy( pucSigPtr, &ucTemp[ 32 ], 32 );
+    }
+    else
+    {
+        pucSigPtr[ 0 ] = 0x20;
+        pucSigPtr++;
+        memcpy( pucSigPtr, &ucTemp[ 32 ], 32 );
+    }
+
+    *pxSigLen = ( CK_ULONG ) pucSig[ 1 ] + 2;
+
+    if( xResult != 0 )
+    {
+        configPRINTF( ( "ERROR: Failure in signing callback: %d \r\n", xResult ) );
+        lFinalResult = TLS_ERROR_SIGN;
+    }
+
+    return lFinalResult;
+}
+
+/*-----------------------------------------------------------*/
+
+/* @brief Provision Device Using Device and CA Certificates
+ *
+ *
+ *	\return    Void
+ *  Prints error message upon failure, success message upon success
+ *
+ */
+static int xProvisionDeviceForJITP( void )
+{
+    /* Provisioning Device Certificate */
+    CK_OBJECT_HANDLE xObject = 0;
+
+    xResult = xProvisionCertificate( xGlobalSession,
+                                     keyCLIENT_CERTIFICATE_PEM,
+                                     sizeof( keyCLIENT_CERTIFICATE_PEM ),
+                                     ( uint8_t * ) pkcs11configLABEL_DEVICE_CERTIFICATE_FOR_TLS,
+                                     &xObject );
+
+    if( ( xResult != CKR_OK ) || ( xObject == CK_INVALID_HANDLE ) )
+    {
+        configPRINTF( ( "ERROR: Failed to provision device certificate. %d \r\n", xResult ) );
+        return -1;
+    }
+
+    /* Provisioning JITR CA Certificate */
+    if( xResult == CKR_OK )
+    {
+        if( sizeof( keyJITR_DEVICE_CERTIFICATE_AUTHORITY_PEM ) != 0 )
+        {
+            xResult = xProvisionCertificate( xGlobalSession,
+                                             keyJITR_DEVICE_CERTIFICATE_AUTHORITY_PEM,
+                                             sizeof( keyJITR_DEVICE_CERTIFICATE_AUTHORITY_PEM ),
+                                             ( uint8_t * ) pkcs11configLABEL_JITP_CERTIFICATE,
+                                             &xObject );
+
+            xResult = CKR_OK;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        configPRINTF( ( "Device credential provisioning succeeded.\r\n" ) );
+        return 0;
+    }
+    else
+    {
+        configPRINTF( ( "ERROR: %d - Device credential provisioning failed.\r\n", xResult ) );
+        return -1;
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+/* @brief Provisions a Device Using Internally Generated Keys
+ *
+ *
+ *  \return Void function, console outputs upon success/failure
+ *
+ */
+static void xDeviceProvisioningForJITP( void )
```

Comment:
I think that there had been a comment at some point about changing the signature of the various "DevModeProvisioning" routines to return status codes. I recommend that change, since it makes it easier for the application writer to isolate problems (say, if they start using a secure element instead).

Reply:
I changed all provisioning functions to return integer error code values. Thank you!